ixfr.c

```diff
@@ -2669,27 +2689,9 @@ static int ixfr_read_one_more_file(struct nsd* nsd, struct zone* zone,
 	const char* zfile, int num_files, uint32_t *dest_serial)
 {
 	char ixfrfile[1024+24];
-	FILE* in;
 	int file_num = num_files+1;
 	make_ixfr_name(ixfrfile, sizeof(ixfrfile), zfile, file_num);
-	in = fopen(ixfrfile, "r");
-	if(!in) {
-		if(errno == ENOENT) {
```

Comment:
Yes, the parser logs errors. The idea is that the application registers a log callback function of type `zone_log_t` in `options.log.callback` that gets called if an error occurs. The function can then format/decorate the error as desired and pass it to it's actual log function. i.e. `log_msg` in the case of NSD. While in the callback, the application can use `zone_log` (or `zone_error`, `zone_warning` or `zone_info`) to log a message using the callback registered by the application. That's how it's supposed to work at least. In this case the parser should print ab error message and return `ZONE_READ_ERROR` (or `ZONE_OUT_OF_MEMORY` in case of `ENOMEM`). I did forget about implementing the printing a message part though :sweat_smile: (so, I'll fix this in an upcoming commit)

Reply:
But the test for ENOENT is necessary for the routine that calls the read of ixfr files. Hence that may need more attention and not just deleting the check, as it goes from a check on ENOENT to a version of the code without a check for ENOENT, and I think that likely means the case may not be handled as it should. I could be nice to handle it by itself anyway for the reason that it is part of the read of ixfr files routine.

Reply:
This is fixed by NLnetLabs/simdzone#185.

Reply:
Apart from the parser making an error callback, the nsd side of the code has to actually handle the file nonexisting case. Because it has to detect that the ixfr file does not exist, at this place or somewhere else in the NSD code that deals with the ixfr files.

Reply:
Can you elaborate? I think I'm overlooking a case(?) `ixfr_read_from_file` repeatedly calls `ixfr_read_one_more_file` until there's no more files. Judging by the code that was previously there, no distinction is made between a file not being there or an error condition(?) Therefore, the operation stays the same? At least, that was the intent. The zone parser reads the file, if it fails, `0` (zero) is returned, otherwise `1`.

Reply:
The code links did not look at pretty as I wanted. So the error code that is returned is fine, and now the parse is fixed it always prints an error. In the old sequence of code, it would not print an error, because it is not an error to find the final ixfr file, and then stop processing the list of ixfr files. Does the new code now print an error every time a list of ixfr files is read?

Reply:
Thanks, I think I get what you're saying. If the file that is passed to `zone_parse` cannot be found, it is considered an error by the parser. In this case, that should not happen because it's expected behavior. To fix that, I can check if the `file` is `NULL` in `ixfr_data_log` and then simply not print the message. `ixfr_data_read` would then return `0`, which would lead to `ixfr_read_from_file` stopping the loop. I think, with the last addition of not printing the message, the behavior will be the same as before?

Reply:
Yes, I think that would work! The code could also call stat() on the file before opening it, is that is more convenient.

Reply:
I've chosen to add a call to `stat`. Thanks :+1: